src/core/queryObserver.ts

```diff
@@ -148,6 +148,37 @@ export class QueryObserver<
     this.currentQuery.removeObserver(this)
   }
 
+  createQueryResult(
+    queryObserverResult: QueryObserverResult<TData, TError>
+  ): QueryObserverResult<TData, TError> {
+    if (this.options.notifyOnChangeTracked) {
+      const trackedResult = {}
```

Comment:
Think we should store the tracked result object on the instance and update it only when the result object changes. This will make sure the reference also stays the same between renders for the tracked version

Reply:
yes, good catch. did that here: 978fb758

i had to adapt the test because capturing the state now led to capturing the same instance, and the get interceptor would return the same data for both captured states. I hope this is just a problem in the way we write these tests...

/edit oof, I think this might lead to stale closure issues when accessing queryInfo in an async way:

consider:

```
useEffect(() => {
  console.log(queryInfo.data)
});
```

this logs `undefined` (while we are loading), then the `data` object, but this:

```
useEffect(() => {
  setTimeout(() => console.log(queryInfo.data), 500);
});
```

will log the data object twice, because at the time we invoke the getter, we have already updated the single instance...

is the referential equality important for the returned queryInfo?

Reply:
aah, I think a better approach might be to create the `trackedResult` new whenever we also create a new `currentResult`, which would be here:

https://github.com/tannerlinsley/react-query/blob/b0c2f6795292d1580e112794d18b27adff77bf48/src/core/queryObserver.ts#L479-L481

That should get rid of the aforementioned issue and also make the code a bit more transparent - writing to currentResult / trackedResult at the same place only. I'll try to do that tomorrow

Reply:
I think I got it: 4ea4eaa3

I'll add tests for referential equality, too.